types/wlr_xdg_foreign_v1.c

```diff
@@ -0,0 +1,491 @@
+		return;
+	}
+	if (wlr_surface_is_xdg_surface(wlr_surface)
+			!= wlr_surface_is_xdg_surface(wlr_surface_child)) {
+		wl_resource_post_error(resource, -1,
+				"surfaces must have the same role");
+		return;
+	}
+	struct wlr_xdg_imported_child_v1 *child;
+	wl_list_for_each(child, &imported->children, link) {
+		if (child->surface == wlr_surface_child) {
+			return;
+		}
+	}
+
+	child = calloc(1, sizeof(struct wlr_xdg_imported_child_v1));
+	if (child == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	child->surface = wlr_surface_child;
+	child->xdg_surface_unmap.notify = handle_child_xdg_surface_unmap;
+	child->xdg_toplevel_set_parent.notify = handle_xdg_toplevel_set_parent;
+
+	if (wlr_surface_is_xdg_surface(wlr_surface)) {
+		struct wlr_xdg_surface *surface =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface);
+		struct wlr_xdg_surface *surface_child =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface_child);
+
+		wlr_xdg_toplevel_set_parent(surface_child, surface);
+		wl_signal_add(&surface_child->events.unmap,
+				&child->xdg_surface_unmap);
+		wl_signal_add(&surface_child->toplevel->events.set_parent,
+				&child->xdg_toplevel_set_parent);
+	} else if (wlr_surface_is_xdg_surface_v6(wlr_surface)) {
+		struct wlr_xdg_surface_v6 *surface =
+			wlr_xdg_surface_v6_from_wlr_surface(wlr_surface);
+		struct wlr_xdg_surface_v6 *surface_child =
+			wlr_xdg_surface_v6_from_wlr_surface(wlr_surface_child);
+
+		wlr_xdg_toplevel_v6_set_parent(surface_child, surface);
+		wl_signal_add(&surface_child->events.unmap,
+				&child->xdg_surface_unmap);
+		wl_signal_add(&surface_child->toplevel->events.set_parent,
+				&child->xdg_toplevel_set_parent);
+	}
+
+	wl_list_insert(&imported->children, &child->link);
+}
+
+static const struct zxdg_imported_v1_interface xdg_imported_impl = {
+	.destroy = xdg_imported_handle_destroy,
+	.set_parent_of = xdg_imported_handle_set_parent_of
+};
+
+static struct wlr_xdg_exported_v1 *xdg_exported_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exported_v1_interface,
+		&xdg_exported_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_exported_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zxdg_exported_v1_interface xdg_exported_impl = {
+	.destroy = xdg_exported_handle_destroy
+};
+
+static void xdg_exporter_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wlr_xdg_exporter_v1 *xdg_exporter_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exporter_v1_interface,
+		&xdg_exporter_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void disconnect_imported(struct wlr_xdg_imported_v1 *imported) {
+	if (imported->exported != NULL) {
+		imported->exported = NULL;
+		zxdg_imported_v1_send_destroyed(imported->resource);
```

Comment:
My understanding is that sending destroyed doesn't necessarily destroy the `imported`. So if the importing client delays the destruction of the `imported` resource, parent-child relationships won't be removed even though they should be.

Reply:
Sorry for the delay. This was refactored, please check to see if it still has the same issue.